akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -147,7 +192,24 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
 
       override protected def logSource: Class[_] = if (isServer) classOf[Http2ServerDemux] else classOf[Http2ClientDemux]
 
-      val multiplexer = createMultiplexer(frameOut, StreamPrioritizer.first())
+      def pushFrameOut(event: FrameEvent): Unit = {
+        pingState match {
+          case OptionVal.None =>
+          case OptionVal.Some(state) =>
+            event match {
+              case _: StreamFrameEvent => state.onDataFrameSeen()
+              case _                   =>
+            }
+        }
+        frameOut.push(event)
+      }
+
+      val multiplexer = createMultiplexer(pushFrameOut, StreamPrioritizer.first())
+      frameOut.setHandler(multiplexer)
+
+      val pingState: OptionVal[ConfigurablePing.PingState] =
```

Comment:
Maybe you could make `PingState` a trait and provide a no-op instance, so we get rid of the extra `OptionVal` stuff?

Reply:
Yeah, good idea, was considering but could not decide.